File: backend/app/services/stats/spending.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import String, and_, case, cast, desc, func, literal, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.models import Endpoint, UsageRequestEvent

from app.services.stats.time_presets import resolve_time_preset
from app.services.user_settings import get_report_currency_preferences


async def get_spending_report(
    db: AsyncSession,
    *,
    profile_id: int,
    preset: str | None = None,
    from_time: datetime | None = None,
    to_time: datetime | None = None,
    api_family: str | None = None,
    model_id: str | None = None,
    endpoint_id: int | None = None,
    connection_id: int | None = None,
    group_by: str = "none",
    limit: int = 50,
    offset: int = 0,
    top_n: int = 5,
) -> dict[str, object]:
    from_time, to_time = resolve_time_preset(preset, from_time, to_time)

    filters = [UsageRequestEvent.profile_id == profile_id]
    if from_time is not None:
        filters.append(UsageRequestEvent.created_at >= from_time)
    if to_time is not None:
        filters.append(UsageRequestEvent.created_at <= to_time)
    if api_family:
        filters.append(UsageRequestEvent.api_family == api_family)
    if model_id:
        filters.append(UsageRequestEvent.model_id == model_id)
    if endpoint_id is not None:
        filters.append(UsageRequestEvent.endpoint_id == endpoint_id)
    if connection_id is not None:
        filters.append(UsageRequestEvent.connection_id == connection_id)
    where = and_(*filters) if filters else literal(True)
    success_where = and_(where, UsageRequestEvent.success_flag.is_(True))
    unpriced_success_where = and_(
        where,
        UsageRequestEvent.success_flag.is_(True),
        or_(
            UsageRequestEvent.priced_flag.is_(False),
            UsageRequestEvent.priced_flag.is_(None),
        ),
    )

    spend_case = case(
        (
            UsageRequestEvent.billable_flag.is_(True),
            func.coalesce(UsageRequestEvent.total_cost_user_currency_micros, 0),
        ),
        else_=0,
    )
    priced_case = case(
        (
            and_(
                UsageRequestEvent.success_flag.is_(True),
                UsageRequestEvent.priced_flag.is_(True),
            ),
            1,
        ),
        else_=0,
    )
    unpriced_case = case(
        (unpriced_success_where, 1),
        else_=0,
    )
    endpoint_label_expr = case(
        (Endpoint.name.is_not(None), Endpoint.name),
        (Endpoint.base_url.is_not(None), Endpoint.base_url),
        (
            UsageRequestEvent.endpoint_id.is_not(None),
            literal("Endpoint ") + cast(UsageRequestEvent.endpoint_id, String),
        ),
        else_=literal("unknown_endpoint"),
    )
    grouped_source = UsageRequestEvent.__table__.outerjoin(
        Endpoint.__table__,
        and_(
            Endpoint.profile_id == UsageRequestEvent.profile_id,
            Endpoint.id == UsageRequestEvent.endpoint_id,
        ),
    )
    summary_row = (
        await db.execute(
            select(
                func.coalesce(func.sum(spend_case), 0).label("total_cost_micros"),
                func.count().label("successful_request_count"),
                func.coalesce(func.sum(priced_case), 0).label("priced_request_count"),
                func.coalesce(func.sum(unpriced_case), 0).label(
                    "unpriced_request_count"
                ),
                func.coalesce(
                    func.sum(func.coalesce(UsageRequestEvent.input_tokens, 0)), 0
                ).label("total_input_tokens"),
                func.coalesce(
                    func.sum(func.coalesce(UsageRequestEvent.output_tokens, 0)), 0
                ).label("total_output_tokens"),
                func.coalesce(
                    func.sum(
                        func.coalesce(UsageRequestEvent.cache_read_input_tokens, 0)
                    ),
                    0,
                ).label("total_cache_read_input_tokens"),
                func.coalesce(
                    func.sum(
                        func.coalesce(UsageRequestEvent.cache_creation_input_tokens, 0)
                    ),
                    0,
                ).label("total_cache_creation_input_tokens"),
                func.coalesce(
                    func.sum(func.coalesce(UsageRequestEvent.reasoning_tokens, 0)), 0
                ).label("total_reasoning_tokens"),
                func.coalesce(
                    func.sum(func.coalesce(UsageRequestEvent.total_tokens, 0)), 0
                ).label("total_tokens"),
            ).where(success_where)
        )
    ).one()

    successful_request_count = int(summary_row.successful_request_count or 0)
    total_cost_micros = int(summary_row.total_cost_micros or 0)
    avg_cost_per_success = (
        int(total_cost_micros / successful_request_count)
        if successful_request_count > 0
        else 0
    )

    group_expr = None
    if group_by == "day":
        group_expr = func.to_char(UsageRequestEvent.created_at, "YYYY-MM-DD")
    elif group_by == "week":
        group_expr = func.to_char(
            func.date_trunc("week", UsageRequestEvent.created_at), 'IYYY-"W"IW'
        )
    elif group_by == "month":
        group_expr = func.to_char(UsageRequestEvent.created_at, "YYYY-MM")
    elif group_by == "api_family":
        group_expr = UsageRequestEvent.api_family
    elif group_by == "model":
        group_expr = UsageRequestEvent.model_id
    elif group_by == "endpoint":
        group_expr = endpoint_label_expr
    elif group_by == "model_endpoint":
        group_expr = (
            cast(UsageRequestEvent.model_id, String)
            + literal("#")
            + func.coalesce(
                cast(UsageRequestEvent.endpoint_id, String),
                literal("-1"),
            )
        )

    groups: list[dict[str, int | str]] = []
    groups_total = 0
    if group_expr is not None:
        groups_total = (
            await db.execute(
                select(func.count()).select_from(
                    select(group_expr.label("group_key"))
                    .select_from(grouped_source)
                    .where(success_where)
                    .group_by(group_expr)
                    .subquery()
                )
            )
        ).scalar_one()

        grouped_rows = (
            await db.execute(
                select(
                    group_expr.label("key"),
                    func.count().label("total_requests"),
                    func.coalesce(func.sum(priced_case), 0).label("priced_requests"),
                    func.coalesce(func.sum(unpriced_case), 0).label(
                        "unpriced_requests"
                    ),
                    func.coalesce(
                        func.sum(func.coalesce(UsageRequestEvent.total_tokens, 0)),
                        0,
                    ).label("total_tokens"),
                    func.coalesce(func.sum(spend_case), 0).label("total_cost_micros"),
                )
                .select_from(grouped_source)
                .where(success_where)
                .group_by(group_expr)
                .order_by(desc(func.coalesce(func.sum(spend_case), 0)))
                .limit(limit)
                .offset(offset)
            )
        ).all()

        groups = [
            {
                "key": row.key or "unknown",
                "total_cost_micros": int(row.total_cost_micros or 0),
                "total_requests": int(row.total_requests or 0),
                "priced_requests": int(row.priced_requests or 0),
                "unpriced_requests": int(row.unpriced_requests or 0),
                "total_tokens": int(row.total_tokens or 0),
            }
            for row in grouped_rows
        ]
    else:
        groups = [
            {
                "key": "all",
                "total_cost_micros": total_cost_micros,
                "total_requests": successful_request_count,
                "priced_requests": int(summary_row.priced_request_count or 0),
                "unpriced_requests": int(summary_row.unpriced_request_count or 0),
                "total_tokens": int(summary_row.total_tokens or 0),
            }
        ]
        groups_total = 1

    top_model_rows = (
        await db.execute(
            select(
                UsageRequestEvent.model_id,
                func.coalesce(func.sum(spend_case), 0).label("total_cost_micros"),
            )
            .where(success_where)
            .group_by(UsageRequestEvent.model_id)
            .having(func.coalesce(func.sum(spend_case), 0) > 0)
            .order_by(desc(func.coalesce(func.sum(spend_case), 0)))
            .limit(top_n)
        )
    ).all()

    top_endpoint_rows = (
        await db.execute(
            select(
                UsageRequestEvent.endpoint_id,
                endpoint_label_expr.label("endpoint_label"),
                func.coalesce(func.sum(spend_case), 0).label("total_cost_micros"),
            )
            .select_from(grouped_source)
            .where(success_where)
            .group_by(UsageRequestEvent.endpoint_id, endpoint_label_expr)
            .having(func.coalesce(func.sum(spend_case), 0) > 0)
            .order_by(desc(func.coalesce(func.sum(spend_case), 0)))
            .limit(top_n)
        )
    ).all()

    unpriced_reason_rows = (
        await db.execute(
            select(
                UsageRequestEvent.unpriced_reason.label("reason"),
                func.count().label("reason_count"),
            )
            .where(unpriced_success_where)
            .group_by("reason")
            .order_by(desc(func.count()))
        )
    ).all()

    (
        report_currency_code,
        report_currency_symbol,
    ) = await get_report_currency_preferences(
        db,
        profile_id=profile_id,
    )

    unpriced_breakdown: dict[str, int] = {}
    for row in unpriced_reason_rows:
        reason = row[0]
        reason_count = row[1]
        key = str(reason or "UNKNOWN")
        unpriced_breakdown[key] = unpriced_breakdown.get(key, 0) + int(
            reason_count or 0
        )

    return {
        "summary": {
            "total_cost_micros": total_cost_micros,
            "successful_request_count": successful_request_count,
            "priced_request_count": int(summary_row.priced_request_count or 0),
            "unpriced_request_count": int(summary_row.unpriced_request_count or 0),
            "total_input_tokens": int(summary_row.total_input_tokens or 0),
            "total_output_tokens": int(summary_row.total_output_tokens or 0),
            "total_cache_read_input_tokens": int(
                summary_row.total_cache_read_input_tokens or 0
            ),
            "total_cache_creation_input_tokens": int(
                summary_row.total_cache_creation_input_tokens or 0
            ),
            "total_reasoning_tokens": int(summary_row.total_reasoning_tokens or 0),
            "total_tokens": int(summary_row.total_tokens or 0),
            "avg_cost_per_successful_request_micros": avg_cost_per_success,
        },
        "groups": groups,
        "groups_total": int(groups_total or 0),
        "top_spending_models": [
            {
                "model_id": row.model_id,
                "total_cost_micros": int(row.total_cost_micros or 0),
            }
            for row in top_model_rows
        ],
        "top_spending_endpoints": [
            {
                "endpoint_id": row.endpoint_id,
                "endpoint_label": row.endpoint_label,
                "total_cost_micros": int(row.total_cost_micros or 0),
            }
            for row in top_endpoint_rows
        ],
        "unpriced_breakdown": unpriced_breakdown,
        "report_currency_code": report_currency_code,
        "report_currency_symbol": report_currency_symbol,
    }

File: backend/internal/domain/stats/spending.go
package stats

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"time"
)

type SpendingParams struct {
	ProfileID          int
	Preset             string
	FromTime           *time.Time
	ToTime             *time.Time
	APIFamily          *string
	IngressModelID     *string
	FinalTargetModelID *string
	EndpointID         *int
	ConnectionID       *int
	GroupBy            string
	Limit              int
	Offset             int
	TopN               int
	ReferenceNow       time.Time
	Scope              string
}

type spendingGroupAggregate struct {
	Key              string
	TotalCostMicros  int64
	TotalRequests    int
	PricedRequests   int
	UnpricedRequests int
	TotalTokens      int
	CostSamples      int
}

func GetSpending(ctx context.Context, exec queryExecutor, params SpendingParams) (SpendingReportResponse, error) {
	scope, err := NormalizeScope(params.Scope)
	if err != nil {
		return SpendingReportResponse{}, err
	}
	groupBy, err := validateSpendingGroupBy(scope, params.GroupBy)
	if err != nil {
		return SpendingReportResponse{}, err
	}
	if scope == ScopeRouteAttempt {
		response := SpendingReportResponse{Groups: []SpendingGroupRow{}, TopSpendingModels: []SpendingTopModel{}, TopSpendingEndpoints: []SpendingTopEndpoint{}, UnpricedBreakdown: map[string]int{}}
		preset := normalizeModelSpendingPreset(params.Preset)
		if params.FromTime != nil && params.ToTime != nil {
			preset = "custom"
		}
		_, coverage, coverageErr := ResolveDatasetCoverage(ctx, exec, "request_logs", preset, params.FromTime, params.ToTime, params.ReferenceNow.UTC())
		if coverageErr != nil {
			return SpendingReportResponse{}, coverageErr
		}
		response.ReportCurrencyCode, response.ReportCurrencySymbol, err = loadReportCurrencyPreferences(ctx, exec, params.ProfileID)
		if err != nil {
			return SpendingReportResponse{}, err
		}
		response.Scope = scope
		response.Caliber = CaliberForScope(scope)
		response.Coverage = DatasetCoverage{RequestLogs: &coverage}
		return response, nil
	}
	limit := params.Limit
	if limit <= 0 {
		limit = 50
	}
	offset := params.Offset
	if offset < 0 {
		offset = 0
	}
	topN := params.TopN
	if topN <= 0 {
		topN = 5
	}
	coveragePreset := normalizeModelSpendingPreset(params.Preset)
	if params.FromTime != nil && params.ToTime != nil {
		coveragePreset = "custom"
	}
	bounds, coverage, err := ResolveDatasetCoverage(ctx, exec, "usage_request_events", coveragePreset, params.FromTime, params.ToTime, params.ReferenceNow.UTC())
	if err != nil {
		return SpendingReportResponse{}, err
	}
	startAt, endAt := &bounds.UsageFrom, &bounds.UsageTo
	records, err := loadUsageEventRecords(ctx, exec, params.ProfileID, startAt, endAt, params.APIFamily, nil, params.EndpointID, params.ConnectionID)
	if err != nil {
		return SpendingReportResponse{}, err
	}
	successRecords := make([]usageEventRecord, 0)
	for _, record := range records {
		if params.IngressModelID != nil && record.ModelID != strings.TrimSpace(*params.IngressModelID) {
			continue
		}
		if params.FinalTargetModelID != nil && (record.ResolvedTargetModelID == nil || *record.ResolvedTargetModelID != strings.TrimSpace(*params.FinalTargetModelID)) {
			continue
		}
		if record.SuccessFlag {
			successRecords = append(successRecords, record)
		}
	}
	response := SpendingReportResponse{Scope: scope, Caliber: CaliberForScope(scope), Coverage: DatasetCoverage{UsageRequestEvents: &coverage}, Groups: []SpendingGroupRow{}, TopSpendingModels: []SpendingTopModel{}, TopSpendingEndpoints: []SpendingTopEndpoint{}, UnpricedBreakdown: map[string]int{}}
	for _, record := range successRecords {
		spend := int64(0)
		if record.TrustedKnownCost() {
			spend = record.TotalCostUserCurrencyMicros
			response.Summary.CostSampleCount++
		} else if record.PricingStatus != "ineligible" {
			response.Summary.CostMissingCount++
		}
		response.Summary.TotalCostMicros += spend
		response.Summary.SuccessfulRequestCount++
		addSpendingSummaryTokenTotals(&response.Summary, record)
		if record.Priced() {
			response.Summary.PricedRequestCount++
		} else {
			response.Summary.UnpricedRequestCount++
			reason := "UNKNOWN"
			if record.UnpricedReason != nil && strings.TrimSpace(*record.UnpricedReason) != "" {
				reason = strings.TrimSpace(*record.UnpricedReason)
			}
			response.UnpricedBreakdown[reason]++
		}
	}
	if response.Summary.SuccessfulRequestCount > 0 {
		response.Summary.AvgCostPerSuccessfulRequestMicros = response.Summary.TotalCostMicros / int64(response.Summary.SuccessfulRequestCount)
	}
	if response.Summary.CostSampleCount > 0 {
		cost := response.Summary.TotalCostMicros
		response.Summary.KnownCostMicros = &cost
	}
	groupAggregates := map[string]*spendingGroupAggregate{}
	modelSpend := map[string]int64{}
	modelCostSamples := map[string]int{}
	modelLabels := map[string]string{}
	endpointSpend := map[string]*SpendingTopEndpoint{}
	for _, record := range successRecords {
		spend := int64(0)
		if record.TrustedKnownCost() {
			spend = record.TotalCostUserCurrencyMicros
		}
		endpointLabel := usageEventEndpointLabel(record)
		groupKey := spendingGroupKey(scope, groupBy, record, endpointLabel)
		groupDisplayKey := groupKey
		if groupBy == "endpoint" {
			groupKey = spendingEndpointKey(record.EndpointID, endpointLabel)
			groupDisplayKey = endpointLabel
		}
		if groupBy == "" || groupBy == "none" {
			groupKey = "all"
			groupDisplayKey = "all"
		}
		aggregate := groupAggregates[groupKey]
		if aggregate == nil {
			groupAggregates[groupKey] = &spendingGroupAggregate{Key: groupDisplayKey}
			aggregate = groupAggregates[groupKey]
		}
		aggregate.TotalCostMicros += spend
		if record.TrustedKnownCost() {
			aggregate.CostSamples++
		}
		aggregate.TotalRequests++
		if record.Priced() {
			aggregate.PricedRequests++
		} else {
			aggregate.UnpricedRequests++
		}
		aggregate.TotalTokens += record.TotalTokens
		modelID := record.ModelID
		if scope == ScopeFinal {
			modelID = "unattributed"
			if record.ResolvedTargetModelID != nil && strings.TrimSpace(*record.ResolvedTargetModelID) != "" {
				modelID = strings.TrimSpace(*record.ResolvedTargetModelID)
			}
		}
		modelLabel := strings.TrimSpace(modelID)
		if scope == ScopeIngress && record.CurrentModelLabel != nil && strings.TrimSpace(*record.CurrentModelLabel) != "" {
			modelLabel = strings.TrimSpace(*record.CurrentModelLabel)
		}
		if existingLabel, ok := modelLabels[modelID]; !ok || existingLabel == strings.TrimSpace(modelID) {
			modelLabels[modelID] = modelLabel
		}
		modelSpend[modelID] += spend
		if record.TrustedKnownCost() {
			modelCostSamples[modelID]++
		}
		endpointKey := spendingEndpointKey(record.EndpointID, endpointLabel)
		endpoint := endpointSpend[endpointKey]
		if endpoint == nil {
			endpointSpend[endpointKey] = &SpendingTopEndpoint{EndpointID: record.EndpointID, EndpointLabel: endpointLabel}
			endpoint = endpointSpend[endpointKey]
		}
		if record.TrustedKnownCost() {
			if endpoint.KnownCostMicros == nil {
				endpoint.KnownCostMicros = new(int64)
			}
			*endpoint.KnownCostMicros += spend
		}
	}
	groupItems := make([]SpendingGroupRow, 0)
	if groupBy == "" || groupBy == "none" {
		groupItems = append(groupItems, spendingGroupRow("all", response.Summary.TotalCostMicros, response.Summary.CostSampleCount, response.Summary.SuccessfulRequestCount, response.Summary.PricedRequestCount, response.Summary.UnpricedRequestCount, response.Summary.TotalTokens))
		response.GroupsTotal = 1
	} else {
		response.GroupsTotal = len(groupAggregates)
		orderedGroups := make([]SpendingGroupRow, 0, len(groupAggregates))
		for _, aggregate := range groupAggregates {
			orderedGroups = append(orderedGroups, spendingGroupRow(aggregate.Key, aggregate.TotalCostMicros, aggregate.CostSamples, aggregate.TotalRequests, aggregate.PricedRequests, aggregate.UnpricedRequests, aggregate.TotalTokens))
		}
		sort.Slice(orderedGroups, func(i int, j int) bool {
			if int64Value(orderedGroups[i].KnownCostMicros) != int64Value(orderedGroups[j].KnownCostMicros) {
				return int64Value(orderedGroups[i].KnownCostMicros) > int64Value(orderedGroups[j].KnownCostMicros)
			}
			return orderedGroups[i].Key < orderedGroups[j].Key
		})
		start := offset
		if start > len(orderedGroups) {
			start = len(orderedGroups)
		}
		end := start + limit
		if end > len(orderedGroups) {
			end = len(orderedGroups)
		}
		groupItems = orderedGroups[start:end]
	}
	response.Groups = groupItems
	modelItems := make([]SpendingTopModel, 0, len(modelSpend))
	for modelID, spend := range modelSpend {
		if modelCostSamples[modelID] == 0 {
			continue
		}
		modelLabel := strings.TrimSpace(modelLabels[modelID])
		if modelLabel == "" {
			modelLabel = strings.TrimSpace(modelID)
		}
		cost := spend
		modelItems = append(modelItems, SpendingTopModel{ModelID: modelID, ModelLabel: modelLabel, KnownCostMicros: &cost})
	}
	sort.Slice(modelItems, func(i int, j int) bool {
		if int64Value(modelItems[i].KnownCostMicros) != int64Value(modelItems[j].KnownCostMicros) {
			return int64Value(modelItems[i].KnownCostMicros) > int64Value(modelItems[j].KnownCostMicros)
		}
		return modelItems[i].ModelID < modelItems[j].ModelID
	})
	if len(modelItems) > topN {
		modelItems = modelItems[:topN]
	}
	response.TopSpendingModels = modelItems
	endpointItems := make([]SpendingTopEndpoint, 0, len(endpointSpend))
	for _, item := range endpointSpend {
		if item.KnownCostMicros == nil {
			continue
		}
		endpointItems = append(endpointItems, *item)
	}
	sort.Slice(endpointItems, func(i int, j int) bool {
		if int64Value(endpointItems[i].KnownCostMicros) != int64Value(endpointItems[j].KnownCostMicros) {
			return int64Value(endpointItems[i].KnownCostMicros) > int64Value(endpointItems[j].KnownCostMicros)
		}
		if endpointIDOrMinusOne(endpointItems[i].EndpointID) != endpointIDOrMinusOne(endpointItems[j].EndpointID) {
			return endpointIDOrMinusOne(endpointItems[i].EndpointID) < endpointIDOrMinusOne(endpointItems[j].EndpointID)
		}
		return endpointItems[i].EndpointLabel < endpointItems[j].EndpointLabel
	})
	if len(endpointItems) > topN {
		endpointItems = endpointItems[:topN]
	}
	response.TopSpendingEndpoints = endpointItems
	response.ReportCurrencyCode, response.ReportCurrencySymbol, err = loadReportCurrencyPreferences(ctx, exec, params.ProfileID)
	if err != nil {
		return SpendingReportResponse{}, err
	}
	return response, nil
}

func addSpendingSummaryTokenTotals(summary *SpendingSummary, record usageEventRecord) {
	summary.TotalInputTokens += record.InputTokens
	summary.TotalOutputTokens += record.OutputTokens
	summary.TotalCacheReadInputTokens += record.CacheReadInputTokens
	summary.TotalCacheCreationInputTokens += record.CacheCreationInputTokens
	summary.TotalReasoningTokens += record.ReasoningTokens
	summary.TotalTokens += record.TotalTokens
}

func spendingGroupKey(scope string, groupBy string, record usageEventRecord, endpointLabel string) string {
	switch groupBy {
	case "day":
		return record.CreatedAt.UTC().Format("2006-01-02")
	case "week":
		year, week := record.CreatedAt.UTC().ISOWeek()
		return fmt.Sprintf("%04d-W%02d", year, week)
	case "month":
		return record.CreatedAt.UTC().Format("2006-01")
	case GroupAPIFamily:
		return record.APIFamily
	case GroupIngressModel:
		return record.ModelID
	case GroupFinalTargetModel:
		if record.ResolvedTargetModelID != nil && strings.TrimSpace(*record.ResolvedTargetModelID) != "" {
			return strings.TrimSpace(*record.ResolvedTargetModelID)
		}
		return "unattributed"
	case GroupEndpoint:
		return endpointLabel
	case GroupTerminalTarget:
		if record.ConnectionID != nil && *record.ConnectionID > 0 {
			return fmt.Sprintf("%d", *record.ConnectionID)
		}
		return "unattributed"
	case "ingress_model_endpoint":
		resolvedEndpointID := -1
		if record.EndpointID != nil {
			resolvedEndpointID = *record.EndpointID
		}
		return fmt.Sprintf("%s#%d", record.ModelID, resolvedEndpointID)
	case "final_target_model_endpoint":
		modelID := "unattributed"
		if record.ResolvedTargetModelID != nil && strings.TrimSpace(*record.ResolvedTargetModelID) != "" {
			modelID = strings.TrimSpace(*record.ResolvedTargetModelID)
		}
		return fmt.Sprintf("%s#%d", modelID, endpointIDOrMinusOne(record.EndpointID))
	default:
		return "all"
	}
}

func validateSpendingGroupBy(scope string, value string) (string, error) {
	normalized := strings.TrimSpace(strings.ToLower(value))
	if normalized == "" {
		normalized = GroupNone
	}
	allowed := map[string]struct{}{GroupNone: {}, "day": {}, "week": {}, "month": {}, GroupAPIFamily: {}}
	if scope == ScopeIngress {
		allowed[GroupIngressModel] = struct{}{}
		allowed["ingress_model_endpoint"] = struct{}{}
	} else if scope == ScopeFinal {
		allowed[GroupFinalTargetModel] = struct{}{}
		allowed[GroupEndpoint] = struct{}{}
		allowed[GroupTerminalTarget] = struct{}{}
		allowed["final_target_model_endpoint"] = struct{}{}
	} else if scope == ScopeRouteAttempt {
		if normalized == GroupNone {
			return normalized, nil
		}
		return "", &HTTPError{StatusCode: 422, Code: "group_invalid", Detail: "route_attempt spending does not support grouping"}
	}
	if _, ok := allowed[normalized]; !ok {
		return "", &HTTPError{StatusCode: 422, Code: "group_invalid", Detail: fmt.Sprintf("group_by %q not allowed for scope %q", value, scope)}
	}
	return normalized, nil
}

func spendingGroupRow(key string, cost int64, costSamples int, requests int, priced int, unpriced int, tokens int) SpendingGroupRow {
	row := SpendingGroupRow{Key: key, TotalCostMicros: cost, CostSampleCount: costSamples, TotalRequests: requests, PricedRequests: priced, UnpricedRequests: unpriced, TotalTokens: tokens}
	if costSamples > 0 {
		value := cost
		row.KnownCostMicros = &value
	}
	return row
}

func spendingEndpointKey(endpointID *int, endpointLabel string) string {
	resolvedEndpointID := -1
	if endpointID != nil {
		resolvedEndpointID = *endpointID
	}
	return fmt.Sprintf("%d\x00%s", resolvedEndpointID, endpointLabel)
}
